### PR TITLE
search: authors exact search

### LINKED ISSUE #1425: author search with and without quotes gives different results
author:A.D.Johnson.1 and author:"A.D.Johnson.1" give different results at the moment. This should not be the case and the behaviour should be the same as in production.

@Panos512 can you have a look?

### PR DESCRIPTION
**Pull request by Panos512**
* Converts `author-search` without double quotes to an `exact-match` search.
  (closes #1425 and #1431)

Signed-off-by: Panos Paparrigopoulos <panos.paparrigopoulos@cern.ch>

### CODE REVIEW

File: inspirehep/modules/search/walkers/elasticsearch.py [hunk 1]
```diff
@@ -118,11 +118,16 @@ def query(keyword):
 
             fields = self.get_fields_for_keyword(keyword, mode='a')
             if fields == ['authors.full_name', 'authors.alternative_name']:
```
Comment (jmartinm) [inline inspirehep/modules/search/walkers/elasticsearch.py]: I am wondering here, is it more robust to check these 2 fields, or better to check the `author` key (from https://github.com/inspirehep/inspire-next/blob/master/inspirehep/config.py#L167).

In that way we can change where the `author` keyword will look but always keep the exact match behaviour.

Comment (kaplun) [inline inspirehep/modules/search/walkers/elasticsearch.py]: This is part of a series of refactoring and design decisions that we need to sit down and make straight. Hopefully part of @jacquerie's future journey in the mapping world.

Meanwhile, :+1: to @jmartinm comment.

File: inspirehep/modules/search/query_factory.py [hunk 2]
```diff
@@ -75,6 +75,7 @@ def invenio_query(pattern, search):
             ))
         finally:
             if current_app.debug:
+                current_app.logger.info(query)
```
Comment (jacquerie) [inline inspirehep/modules/search/query_factory.py]: Isn't one of these two lines redundant?

File: inspirehep/modules/search/walkers/elasticsearch.py [hunk 3]
```diff
@@ -117,12 +119,17 @@ def query(keyword):
...
-                    Q("match", authors__name_variations=str(node.value)),
-                    Q("match", authors__full_name=str(node.value)),
-                    Q("match", authors__inspire_bai=str(node.value))
-                ])
+            if fields == current_app.config['SEARCH_ELASTIC_KEYWORD_MAPPING']['author']:
+                return Q(
+                    'bool',
+                    must=Q('bool', should=[
+                        Q("match", authors__name_variations=str(node.value)),
+                        Q("term", authors__inspire_bai=str(node.value))
```
Comment (jacquerie) [inline inspirehep/modules/search/walkers/elasticsearch.py]: This no longer works: BAIs were moved to the `ids` field.

File: inspirehep/modules/search/query_factory.py (hunk 2 above)
Comment (kaplun) [inline inspirehep/modules/search/query_factory.py]: shouldn't it be `current_app.logger.debug(query)`?

Comment (jacquerie) [inline inspirehep/modules/search/query_factory.py]: Ah, yeah, the log level should definitely be `debug`.

File: inspirehep/modules/search/walkers/elasticsearch.py (hunk 3 above)
Comment (Panos512) [inline inspirehep/modules/search/walkers/elasticsearch.py]: so, should we make `ids` not analyzed in order to be able to match them with `term` queries? 

Also we don't need to copy to `author` and `exactauthor` anymore?

Comment (kaplun) [inline inspirehep/modules/search/walkers/elasticsearch.py]: IDS indeed should never be analyzed. They have to be taken as a literal string.

File: inspirehep/modules/records/mappings/records/hep.json [hunk 4]
```diff
@@ -111,13 +111,13 @@
...
-                            ],
-                            "index": "not_analyzed",
-                            "type": "string"
+                        "ids":{
+                          "properties": {
+                            "value": {
+                              "index": "not_analyzed",
+                              "type": "string"
+                            }
+                          }
```
Comment (jacquerie) [inline inspirehep/modules/records/mappings/records/hep.json]: This change should be in its own commit.

Comment (jacquerie) [inline inspirehep/modules/records/mappings/records/hep.json]: Oh, and run `scripts/prettifyjson`.

File: inspirehep/modules/search/query_factory.py [hunk 5]
```diff
@@ -75,7 +75,7 @@ def invenio_query(pattern, search):
             ))
         finally:
             if current_app.debug:
-                current_app.logger.info(json.dumps(query.to_dict(), indent=4))
+                current_app.logger.debug(json.dumps(query.to_dict(), indent=4))
```
Comment (jacquerie) [inline inspirehep/modules/search/query_factory.py]: Not relevant for this PR, but please read this writeup by @mihaibivol on how to do proper logging: https://github.com/inspirehep/inspire-next/issues/1371.

File: inspirehep/modules/search/walkers/elasticsearch.py [hunk 6]
```diff
@@ -27,6 +27,8 @@
 
 from elasticsearch_dsl import Q
```
Comment (jacquerie) [inline inspirehep/modules/search/walkers/elasticsearch.py]: No space needed here.

File: inspirehep/modules/search/walkers/elasticsearch.py [hunk 7]
```diff
@@ -27,6 +27,8 @@
```
Comment (jacquerie) [inline inspirehep/modules/search/walkers/elasticsearch.py]: It would be cool to add a preparatory commit that adds the `__future__` block and fixes the license. See: https://github.com/inspirehep/inspire-next/pull/1500/commits/e399aa155af1131c222c54909e671df630fe3d4d

**Comment by jacquerie** [comment]
:+1:, it's almost there.

**Comment by Panos512** [comment]
@jacquerie done! 👯
